src/workers/notifications/articleUpvoteMilestone.ts

```diff
@@ -0,0 +1,48 @@
+import { messageToJson } from '../worker';
+import { Post, Upvote } from '../../entity';
+import {
+  NotificationPostContext,
+  NotificationUpvotersContext,
+} from '../../notifications';
+import { NotificationWorker } from './worker';
+import { uniquePostOwners, UPVOTE_MILESTONES } from './utils';
+
+interface Data {
+  userId: string;
+  postId: string;
+}
+
+const worker: NotificationWorker = {
+  subscription: 'api.article-upvote-milestone-notification',
+  handler: async (message, con) => {
+    const data: Data = messageToJson(message);
+    const post = await con
+      .getRepository(Post)
+      .findOne({ where: { id: data.postId } });
+    const users = uniquePostOwners(post, data.userId);
+    if (!users.length || !UPVOTE_MILESTONES.includes(post.upvotes)) {
+      return;
+    }
+    const upvotes = await con.getRepository(Upvote).find({
+      where: { postId: post.id },
+      take: 5,
+      order: { createdAt: 'desc' },
+      relations: ['user'],
+    });
+    const upvoters = await Promise.all(upvotes.map((upvote) => upvote.user));
+    const ctx: Omit<
+      NotificationPostContext & NotificationUpvotersContext,
+      'userId'
+    > = {
+      post,
+      upvoters,
+      upvotes: post.upvotes,
+    };
+    return users.map((userId) => ({
+      type: 'article_upvote_milestone',
+      ctx: { ...ctx, userId },
+    }));
```

Comment:
this one I have some questions about.

Will this not float already shown upvoters in certain cases?
so upvote #5 = create with 5, which would be fine.

Then with 6th it would simply return the same no? (with 1 difference)

Mainly raising this how it would handle on the notification centre with a lot of upvotes on the same article.

Reply:
These are the upvote milestones: `export const UPVOTE_MILESTONES = [
  1, 2, 3, 4, 5, 10, 20, 50, 100, 200, 500, 1000, 1250, 1500, 5000, 7500, 10000,
];`
We show up to 5 latest upvoters. Does that make sense?

Reply:
I do think so, but to clarify for the sake of example let's say I have 1 article and all notifications are upvotes on this article, I would see the following notifications in my screen right?

[ upvoted by user 6, 7, 8, 9, 10 ]
[ upvoted by user 1, 2, 3, 4, 5 ]
[ upvoted by user 1, 2, 3, 4 ]
[ upvoted by user 1, 2, 3 ]
[ upvoted by user 1, 2 ]
[ upvoted by user 1 ]

Have we considered any threshold for when it gets invoked or is it just realtime upvotes?

Reply:
Do you mean like think about throttling it or so?

Reply:
Yeah mainly the ones from 1-5 seems a bit redundant if they would happen right after each other? (so maybe there could be a x minute processing time for certain ones?

Reply:
I think we should check if it's possible to throttle emails and push from the provider side of things. From notification center point of view it kinda makes sense to have them I think. We can also remove the 2,4 milestones. So it will be 1,3,5. Let me know what you think

Reply:
Yeah that makes a lot more sense considering how we query, so let's go for those steps.